File: Top Interview Questions/160. Intersection of Two Linked Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if (headA == null || headB == null) return null;
        
        ListNode a = new ListNode(0);
        ListNode b = new ListNode(0);
        a.next = headA;
        b.next = headB;
        
        while (a.next != b.next) {
            if (a.next != null) {
                a.next = a.next.next;
            } else {
                a.next = headB;
            }
            
            if (b.next != null) {
                b.next = b.next.next;
            } else {
                b.next = headA;
            }
        }
        
        return a.next;
    }
}

File: Top Interview Questions/Two Pointers/160. Intersection of Two Linked Lists.py
# 160. Intersection of Two Linked Lists
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        if not headA or not headB:
            return None

        a, b = ListNode(0), ListNode(0)
        a.next, b.next = headA, headB

        while a.next != b.next:
            if a.next:
                a.next = a.next.next
            else:
                a.next = headB
            if b.next:
                b.next = b.next.next
            else:
                b.next = headA

        return a.next